lib/pkcs11h-core.c

```diff
@@ -1318,6 +1319,10 @@ __pkcs11h_forkFixup (
 		mutex_locked = TRUE;
 #endif
 
+		/* We are already running, so on fork() every possible new child will
+		 * safely finalize the fixup */
+		pkcs11h_setForkMode (FALSE);
+
```

Comment:
What do you mean by "already running"? what do you do with the other providers? Once again, the spec implies that for *every* fork all providers' C_Initialized must be called, this means that in OpenSC you create a loop in which C_Initialize calls fork and in this fork C_Initialize should be called and so on.

Reply:
When we are between the region where I set
```
pkcs11h_setForkMode (FALSE);
```
to finnally
```
pkcs11h_setForkMode (TRUE);
```
we know that no matter how many forks occur inside this block, we are calling `C_Initialize()` for *every* provider exactly once per child. Even if a provider may fork during `C_Initialize()` (e.g. OpenSC), we have already called `C_Initialize()` for every provider before `current` and we will call `C_Initialize()` for every `current->next` in every child. So there's no need to fixup any more forks in this block.